k2/csrc/tensor_ops.cu

```diff
@@ -288,13 +291,96 @@ Tensor Index(Tensor &src, Array1<int32_t> &indexes,
   }
 }
 
+template <typename T>
+static void IndexAdd1DImpl(ContextPtr context, const T *src_data,
+                           int32_t src_dim, const int32_t *indexes_data,
+                           int32_t dest_dim, T *dest_data) {
+  // Note: we do not use static_assert here since T may be double
+  // at compile time.
+  //
+  // This function will never be called with T==double
+  // at runtime.
+  K2_CHECK_EQ(sizeof(T), sizeof(uint32_t));
+  K2_CHECK_EQ(context->GetDeviceType(), kCuda);
+  int32_t num_buckets = dest_dim;
+  int32_t num_elements = src_dim;
+
+  Array1<int32_t> row_splits(context, num_buckets + 1);
+  Array1<int32_t> histogram = row_splits.Range(0, num_buckets);
```

Comment:
As MultiSplit will change both keys and values in place, what I thought originally is that we can re-arrange `index`  use MultiSplit and use the result as the row-ids of the `ragged` below, and gather `ragged.values` with the re-arraged `keys` from `src`, at least for the case where `allow_minus_one=false`

Reply:
For the case where `allow_minus_one=true`, I think we can get the end position of valid row_ids (note I suppose you put -1 in the last bucket) with only one kenerl, say, the previous value in the re-arranged `index` is not -1 but current value is -1.

Reply:
thanks, will try it tomorrow.